proposals/0025-shader-execution-reordering.md

```diff
@@ -0,0 +1,1460 @@
+calling `ReorderThread`. These are valid use cases as reordering can be
+beneficial even when shading happens inline in the raygeneration shader, and
+reordering before a known to be coherent or cheap shader can be
+counterproductive. For cases in which both is desired, keeping `ReorderThread`
+and `HitObject::Invoke` separated is still beneficial as detailed below.
+
+Common hit processing can happen in the raygeneration shader with the
+additional efficiency gains of hit coherence. Benefits include:
+- Logic otherwise duplicated can be hoisted into the raygeneration shader
+without a loss of hit coherence. This can reduce instruction cache pressure
+and reduce compile times.
+- Logic between `ReorderThread` and `HitObject::Invoke` have access to the
+full state of the raygeneration shader. It can access a large material stack
+keeping track of surface boundaries, for example. This is difficult or
+impossible to communicate through the payload.
+- Modularity between the shader stages is improved by allowing hit shading to
+focus on surface logic. Different raygeneration shaders, and different call
+sites within a raygeneration shader, may want to vary how common logic is
+implemented. E.g., on first bounce a shadow ray may be fired after performing
+common light sampling. On a second bounce a shadow map lookup may be enough.
+
+In addition to the above, API complexity is reduced by only having separate
+calls, as opposed to both separate calls and a fused variant. Further,
+`ReorderThread` naturally communicates a reorder point, when hit-coherent
+execution starts and that it will persist after the call (until the next
+reorder point). Reasoning about the execution and that it is hit-coherent is
+not as obvious after a call to a hypothetical (fused)
+`HitObject::ReorderAndInvoke`. Finally, tools can report live state across
+`ReorderThread` and users can optimize live state across it. This is important
+as live state across `ReorderThread` may be more expensive on some
+architectures.
+
+Some examples follow.
+
+### Example: Common computations that rely on large raygeneration state
+
+```C++
+struct IorData
+{
+    float ior;
+    uint id;
+};
+
+IorData iorList[16];
+uint iorListSize = 0;
+
+for( ... )
+{
+    HitObject hit = HitObject::TraceRay( ... );
+    ReorderThread( hit );
+
+    IorData newEntry = LoadIorDataFromHit( hit );
+    bool enter = hit.GetHitKind() == HIT_KIND_TRIANGLE_FRONT_FACE;
+
+    payload.ior = UpdateIorList( newEntry, enter, iorList, iorListSize );
+
+    HitObject::Invoke( hit, payload );
+}
+```
+
+### Example: Do common computations with hit-coherence
+
+```C++
+hit = HitObject::TraceRay( ... );
+ReorderThread( hit );
+
+payload.giData = GlobalIlluminationCacheLookup( hit );
+
+HitObject::Invoke( hit, payload );
+```
+
+### Example: Same surface shader but different behavior in the raygeneration shader
+
+```C++
+// Primary ray wants perfect shadows and does not need to reorder as it is
+// coherent enough.
```

Comment:
We'd appreciate using language that doesn't suggest that there is no reorder at trace ray.
It should be the expectation that impl should only minimize this. 
// Primary ray wants perfect shadows and does not need to reorder more than required and best performant for given implementation
// The domain of shadow rays is expected to be coherent enough.

Reply:
The latest commit includes clarifications like this.